File: leetcode_1008/cpp/leetcode_1008.cpp
//
// Created by wzy on 2022/4/8.
//

#include<iostream>
#include<queue>
#include<vector>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <list>
#include <deque>
#include <stack>
#include <sstream>
#include <typeinfo>
#include <climits>
#include <random>
#include <ctime>

using namespace std;

struct TreeNode{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0),left(nullptr),right(nullptr) {}
    TreeNode(int x) : val(x),left(nullptr),right(nullptr) {}
    TreeNode(int x,TreeNode *left,TreeNode* right):val(x),left(left),right(right) {}
};

class Solution {
public:
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        //根据BST树的性质，分割左右子树区间
        TreeNode* root = builderBST(preorder,0,preorder.size()-1);
        return root;
    }
    vector<string> printTree(TreeNode* root){
        string s;
        vector<string> levelOrder = levelOrderStr(root);
        return levelOrder;
    }
    void deleteTree(TreeNode* root){
        if(root){
            deleteTree(root->left);
            deleteTree(root->right);
            root = nullptr;
        }
    }
    vector<string> levelOrderStr(TreeNode* root){
        vector<TreeNode*> cur;
        vector<TreeNode*> next;
        if(!root)
            return vector<string> {};
        if(root)
            cur.push_back(root);
        vector<string> res;
//        vector<string> curEles;
        while(cur.size()){
            //标识next是否都为null
            int flag = 0;
            for(int i = 0;i < cur.size();i++){
                TreeNode* node = cur[i];
                if(!node){
                    res.push_back("null");
                    continue;
                }
                else
                    res.push_back(to_string(node->val));
                //如果node的孩子不为null,则改变flag
                if(node->left || node->right)
                    flag++;
                next.push_back(node->left);
                next.push_back(node->right);
            }
            cur = next;
            if(flag == 0)
                cur.clear();
            next.clear();
        }
        return res;
    }
private:
    TreeNode* builderBST(vector<int>& preorder,int start,int end){
        if(start > end)
            return nullptr;
        else if(start == end)
            return new TreeNode(preorder[start]);
        TreeNode* root = new TreeNode(preorder[start]);
        int target = root->val;
        //右子树起始索引
        int idx = start + 1;
        while(idx <= end){
            if(preorder[idx] > target)
                break;
            idx++;
        }
        root->left = builderBST(preorder,start+1,idx-1);
        root->right = builderBST(preorder,idx,end);
        return root;
    }
};

template<typename T>
string convertVecToStr(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    return res;
}

template<typename T>
void printVecArr(std::vector<std::vector<T>> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << convertVecToStr(ele);
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

template<typename T>
void printVec(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

int main() {
//    TreeNode *root = new TreeNode(8);
//    TreeNode *left = new TreeNode(5);
//    TreeNode *right = new TreeNode(10);
//    TreeNode *ll = new TreeNode(1);
//    TreeNode *lr = new TreeNode(7);
//    TreeNode *rl = nullptr;
//    TreeNode *rr = new TreeNode(12);
//    root->left = left;
//    root->right = right;
//    left->left = ll;
//    left->right = lr;
//    right->left = rl;
//    right->right = rr;
//    ll->left = new TreeNode(0);
//    lr->right = new TreeNode(5);

//    [8,5,1,7,10,12]
    vector<int> preorder = {8,5,1,7,10,12};
    TreeNode* root = Solution().bstFromPreorder(preorder);
    vector<string> rootArr = Solution().levelOrderStr(root);
    printVec(rootArr);
    //result
    //[8,5,10,1,7,null,12]
}

File: leetcode_1008/java/leetcode_1008.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public class leetcode_1008 {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    class Solution {
        public TreeNode bstFromPreorder(int[] preorder) {
            //根据BST树的性质，分割左右子树区间
            return builderBST(preorder,0,preorder.length-1);
        }
        private TreeNode builderBST(int[] preorder,int start,int end){
            if(start > end)
                return null;
            else if(start == end)
                return new TreeNode(preorder[start]);
            TreeNode root = new TreeNode(preorder[start]);
            int target = root.val;
            //右子树起始索引
            int idx = start + 1;
            while(idx <= end){
                if(preorder[idx] > target)
                    break;
                idx++;
            }
            root.left = builderBST(preorder,start+1,idx-1);
            root.right = builderBST(preorder,idx,end);
            return root;
        }
        public List<List<Integer>> levelOrder(TreeNode root) {
            List<TreeNode> cur = new ArrayList();
            List<TreeNode> next = new ArrayList();
            if (root == null)
                return new LinkedList();
            if (root != null)
                cur.add(root);
            List<List<Integer>> res = new LinkedList();
            List<Integer> curEles = new LinkedList();
            while (cur.size() > 0) {
                for (int i = 0; i < cur.size(); i++) {
                    curEles.add(cur.get(i).val);
                    if (cur.get(i).left != null)
                        next.add(cur.get(i).left);
                    if (cur.get(i).right != null)
                        next.add(cur.get(i).right);
                }
                res.add(new LinkedList(curEles));
                cur = new LinkedList(next);
                curEles.clear();
                next.clear();
            }
            return res;
        }

        public List<String> levelOrderStr(TreeNode root) {
            List<TreeNode> cur = new ArrayList();
            List<TreeNode> next = new ArrayList();
            if (root == null)
                return new LinkedList();
            if (root != null)
                cur.add(root);
            List<String> res = new LinkedList();
            while (cur.size() > 0) {
                int flag = 0;
                for (int i = 0; i < cur.size(); i++) {
                    if (cur.get(i) == null) {
                        res.add("null");
                        continue;
                    } else
                        res.add(String.valueOf(cur.get(i).val));
                    TreeNode left = cur.get(i).left;
                    TreeNode right = cur.get(i).right;
                    if (left != null || right != null)
                        flag++;
                    next.add(left);
                    next.add(right);
                }
                cur = new LinkedList(next);
                if (flag == 0)
                    cur.clear();
                next.clear();
            }
            return res;
        }

        public <T> void printListArr(List<List<T>> list) {
            for (int i = 0; i < list.size(); i++) {
                if (i == list.size() - 1) {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.println(Arrays.toString(list.get(i).toArray(Integer[]::new)) + " ] ");
                } else {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.printf(Arrays.toString(list.get(i).toArray(Integer[]::new)) + " , ");
                }
            }
        }
    }
    public static void main(String[] args) {
        leetcode_1008.Solution solution = new leetcode_1008().new Solution();
//        leetcode_1008.TreeNode root = new leetcode_1008().new TreeNode(10);
//        leetcode_1008.TreeNode left = new leetcode_1008().new TreeNode(5);
//        leetcode_1008.TreeNode right = new leetcode_1008().new TreeNode(15);
//        leetcode_1008.TreeNode ll = new leetcode_1008().new TreeNode(3);
//        leetcode_1008.TreeNode lr = new leetcode_1008().new TreeNode(7);
//        leetcode_1008.TreeNode rl = null;
//        leetcode_1008.TreeNode rr = new leetcode_1008().new TreeNode(18);
//        left.left = ll;
//        left.right = lr;
//        right.left = rl;
//        right.right = rr;
//        root.left = left;
//        root.right = right;
//        ll.left = new leetcode_1008().new TreeNode(0);
//        lr.left = new leetcode_1008().new TreeNode(3);
//        lr.left = new leetcode_1008().new TreeNode(5);
        int[] preorder = {8,5,1,7,10,12};
        //[8,5,1,7,10,12]
        TreeNode root = solution.bstFromPreorder(preorder);
        List<String> rootStr = solution.levelOrderStr(root);
        System.out.println(Arrays.toString(rootStr.toArray(String[]::new)));
        //result
        //[8, 5, 10, 1, 7, null, 12]
    }
}
